pkg/controller/core.oam.dev/v1alpha2/application/revision.go

```diff
@@ -0,0 +1,156 @@
+	"github.com/oam-dev/kubevela/pkg/appfile"
+	"github.com/oam-dev/kubevela/pkg/controller/utils"
+)
+
+// GenerateRevision will generate revision for an Application when created/updated
+func (r *Reconciler) GenerateRevision(ctx context.Context, appfile *appfile.Appfile, application *v1alpha2.Application, ac *v1alpha2.ApplicationConfiguration, comps []*v1alpha2.Component) (bool, *v1alpha2.AppRevisionHash, error) {
+	appHash, err := utils.ComputeSpecHash(application.Spec)
+	if err != nil {
+		return false, nil, err
+	}
+	var componentDefinitionHash, traitDefinitionHash []v1alpha2.NameAndHashValue
+
+	copiedApp := application.DeepCopy()
+	// clean up the status of application going to be recorded in revision
+	copiedApp.Status = v1alpha2.AppStatus{}
+	appRev := &v1alpha2.ApplicationRevision{
+		Spec: v1alpha2.ApplicationRevisionSpec{
+			Application:              copiedApp,
+			ComponentDefinitions:     nil, // TODO: add componentDefinition when completed.
+			Components:               comps,
+			ApplicationConfiguration: ac,
+		},
+	}
+	appRev.Name = appfile.RevisionName
+	appRev.Namespace = application.Namespace
+
+	for _, w := range appfile.Workloads {
+		if w == nil {
+			continue
+		}
+		if !DupWorkloadDef(appRev, w.FullTemplate.WorkloadDefinition) {
+			// TODO(wonderflow): the workloadDefinition is going to refactor to componentDefinition, so we record it as componentDefinition here
+			hashValue, err := utils.ComputeSpecHash(w.FullTemplate.WorkloadDefinition.Spec.DeepCopy())
+			if err != nil {
+				return false, nil, errors.Wrapf(err, "fail to compute hash for workloadDefinition %s", w.FullTemplate.WorkloadDefinition.Name)
+			}
+			componentDefinitionHash = append(componentDefinitionHash, v1alpha2.NameAndHashValue{
+				Name:      w.FullTemplate.WorkloadDefinition.Name,
+				HashValue: hashValue,
+			})
+			appRev.Spec.WorkloadDefinitions = append(appRev.Spec.WorkloadDefinitions, w.FullTemplate.WorkloadDefinition)
+		}
+		for _, t := range w.Traits {
+			if t == nil {
+				continue
+			}
+			if !DupTraitDef(appRev, t.FullTemplate.TraitDefinition) {
+				hashValue, err := utils.ComputeSpecHash(t.FullTemplate.TraitDefinition.Spec.DeepCopy())
+				if err != nil {
+					return false, nil, errors.Wrapf(err, "fail to compute hash for traitDefinition %s", t.FullTemplate.TraitDefinition.Name)
+				}
+				traitDefinitionHash = append(traitDefinitionHash, v1alpha2.NameAndHashValue{
+					Name:      t.FullTemplate.TraitDefinition.Name,
+					HashValue: hashValue,
+				})
+				appRev.Spec.TraitDefinitions = append(appRev.Spec.TraitDefinitions, t.FullTemplate.TraitDefinition)
+			}
+		}
+	}
+	newRevision := &v1alpha2.AppRevisionHash{
+		ApplicationSpec:      appHash,
+		ComponentDefinitions: componentDefinitionHash,
+		TraitDefinitions:     traitDefinitionHash,
+	}
+	// check API spec difference
+	if application.Status.LatestRevision != nil && application.Status.LatestRevision.AppRevisionHash != nil {
```

Comment:
we should compare the newRevision and the latest AppRevision before doing any deepEqual comparison. The fact that the Revision Hash is a structure makes it not very intuitive (and might even be incorrect)